### ISSUE TITLE
Expose parts of Neovim API written in C to vim.api.nvim_get_runtime_files

### ISSUE THREAD

**Issue by theKnightsOfRohan** [neovim/neovim#27069]
### Problem

I've configured `lua_ls` to include the vim runtime files, so I can get LSP support for the vim API and my plugins, with the config shown below:
```lua
require("lspconfig").lua_ls.setup({
    settings = {
        Lua = {
            runtime = {
                -- Tell the language server which version of Lua you're using (most likely LuaJIT in the case of Neovim)
                version = "LuaJIT",
            },
            diagnostics = {
                -- Get the language server to recognize the `vim` global
                globals = { "vim" },
            },
            workspace = {
                -- Make the server aware of Neovim runtime files
                library = vim.api.nvim_get_runtime_file("", true),
            },
            -- Do not send telemetry data containing a randomized but unique identifier
            telemetry = {
                enable = false,
            },
        },
    },
})
```
However, this inclusion is incomplete. I can see that through the below, where the very function I am calling is not a member of the object I am using:
<img width="729" alt="Screenshot 2024-01-17 at 10 29 36 AM" src="https://github.com/neovim/neovim/assets/114779675/e6fba11d-7314-4142-9092-3919526ec92e">

I can go to definition to various parts of the neovim source code, and can even go to references of various functions. However, I can't go to the definition of the functions that are not in the definition, which leads me to believe that these are parts of the neovim API written in C. I want more complete annotations and LSP definitions for the neovim API, as using `:help` can become tedious at times.

So, is there some way to create some sort of header file, written in lua, which can be fed to the `vim.api.nvim_get_runtime_file` function, to get lsp definitions and annotations for the C-specific functions as well?

### Expected behavior

The files fed to the LSP library by the `vim.api.nvim_get_runtime_file` should include some definition for Neovim lua modules written in C. This would make it easier for people to use the neovim API, which can only be a good thing. Without knowing too much about embedded programming and Neovim in general, I think that this shouldn't be TOO hard to implement, as you're annotating preexisting functions in some way.

**Comment by wookayin** [comment]
This is already done in #24523, available in Nvim 0.10-nightly. On Nvim 0.9, you can use some external plugins https://github.com/folke/neodev.nvim (need some configuration)

<img width="1542" alt="image" src="https://github.com/neovim/neovim/assets/1009873/4b9a3a9c-7f54-42e9-b66e-c7b4771ddbae">

**Comment by theKnightsOfRohan** [comment]
@wookayin Okay, thanks!
